apps/backend/src/groups/groups.service.ts

```diff
@@ -93,4 +112,38 @@ export class GroupsService {
 
     return groupToDelete;
   }
+
+  syncUserGroups(user: User, groups: Array<string>) {
+    // Check if user is any existing groups that they should not be
+    user.$get('groups', {include: [User]}).then((currentGroups) => {
+      currentGroups.filter(group => !groups.includes(group.name)).forEach((groupToLeave) => {
+        this.removeUserFromGroup(groupToLeave, user);
+      })
+    })
+
+    groups.forEach((group) => {
+      this.findByName(group).then((existingGroup) => {
+        // Check if the user is already in that group
+        user.$get('groups', {include: [User]}).then((groups) => {
+          const groupMap = groups.map((group) => new GroupDto(group));
+
+          if(!groupMap.includes(existingGroup)) {
+            this.addUserToGroup(existingGroup, user, "member");
+          }
+        });
+      }).catch((err) => {
+        if(err instanceof NotFoundException) {
+
+          const createGroup: CreateGroupDto = {
+            name: group,
+            public: false
+          };
+
+          this.create(createGroup).then((newGroup) => {
+            this.addUserToGroup(newGroup, user, "member");
```

Comment:
would prefer to use a 'guard clause' type of control flow where we get the list of groups that the user is in, create any groups that don't already exist, and then go through and add the user to any groups they are not already in